### ISSUE TITLE
HDF5 engine not writing single value attributes from data array of size = 1

### ISSUE THREAD

**Issue by williamfgc** [ornladios/ADIOS2#1148]
Only attributes of type string work when writing single value attributes from Python bindings using numpy. All numerical supported types are not written.

**Comment by guj** [comment]
@williamfgc 
Hi William,  I just tried, works for  me:
GROUP "/" {
   ATTRIBUTE "dimension-num" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 4
      }
   }
   ATTRIBUTE "dimension0" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 10
      }
   }
   ATTRIBUTE "dimension1" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 20
      }
   }
   ATTRIBUTE "dimension2" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 30
      }
   }
.....

**Comment by williamfgc** [comment]
@guj does it work with bpls?

**Comment by guj** [comment]
@williamfgc  
Yes.

ls -lt ../build/bin/bpls 
-rwxr-xr-x  1 junmin  staff  585940 Feb  7 17:07 ../build/bin/bpls


../build/bin/bpls -a myVector.h5  
  long long int  dimension-num   attr
  long long int  dimension0      attr
  long long int  dimension1      attr
  long long int  dimension2      attr
  long long int  dimension3      attr
  float          h5Floats        {10}
  int            h5Ints          {10}
  double         h5ScalarDouble  scalar
  string         type            attr
  string         version_major   attr
  string         version_minor   attr

**Comment by williamfgc** [comment]
@guj are you using `DefineAttribute<T>(name, value)`  or `DefineAttribute<T>(name, data, elements)` where elements = 1 ?  I am trying the former from a test in the high-level API and it does write things other than string.

**Comment by guj** [comment]
See: examples/hello/hdf5Writer/helloHDF5Writer.cpp 
it is using DefineAttribute<T>(name, value)

**Comment by williamfgc** [comment]
@guj can we make this into a test?

**Comment by williamfgc** [comment]
@guj the hdf5 and bp outputs are completely different in your example, there is something going on with the naming of attributes in hdf5:
./bpls myVector.bp -lav
```
File info:
  of variables:  3
  of attributes: 8
  statistics:    Min / Max 
  string    /adios2_schema/mesh/type          attr   = "explicit"
  long int  adios2_schema/mesh/dimension-num  attr   = 4
  long int  adios2_schema/mesh/dimension0     attr   = 10
  long int  adios2_schema/mesh/dimension1     attr   = 20
  long int  adios2_schema/mesh/dimension2     attr   = 30
  long int  adios2_schema/mesh/dimension3     attr   = 40
  string    adios2_schema/version_major       attr   = "2"
  string    adios2_schema/version_minor       attr   = "3"
  float     h5Floats                          {10} = 0 / 9
  int       h5Ints                            {10} = -9 / 0
  double    h5ScalarDouble                    scalar = 1.234 / 1.234
```

./bpls myVector.h5 -lav
```
File info:
  of variables:  3
  of attributes: 8
  statistics:    Min / Max 

  long int  dimension-num   attr   = 4
  long int  dimension0      attr   = 10
  long int  dimension1      attr   = 20
  long int  dimension2      attr   = 30
  long int  dimension3      attr   = 40
  float     h5Floats        {10} = 0 / 0
  int       h5Ints          {10} = 0 / 0
  double    h5ScalarDouble  scalar = 0 / 0
  string    type            attr   = "explicit"
  string    version_major   attr   = "2"
  string    version_minor   attr   = "3"
```

This is why it's important to make these example into parameterized tests across engines.

**Comment by williamfgc** [comment]
@guj, I see what the problem is. The HDF5 engine is not writing attribute arrays with 1 element. In Python everything comes from arrays (numpy arrays). I just added:

```
std::vector<std::int64_t> array1(1, 10);
        hdf5IO.DefineAttribute<std::int64_t>("array1", array1.data(),
                                             array1.size());
```
to reproduce. Please make sure we have tests.

**Comment by guj** [comment]
Hi William,

Yes, the 1 element was my careless mistake. should have checked >=1 but I did > 1 instead.
Thanks for finding that.

I will update and add array attributes  to the one of the hdf5 test.

The Attribute that has slashes has hierarchical meaning in hdf5. 
e.g. attibute “/a/b” implies group a exists. Otherwise hdf5 exits with error. 

So if path does not exist, I use the value “c” as attire name instead of throw an error. 

This is a temporary simple workaround.

- Junmin

**Comment by pnorbert** [comment]
You support variables with paths by creating the groups recursively, don't
you? Can you do the same with attributes?
Another question: is attribute attached to a variable supported?

**Comment by williamfgc** [comment]
@guj so there are two issues here to be addressed. Tests are ready to go on current master (I turned it off). Please make sure the following HDF5 Python high-level API test runs correctly: 

https://github.com/ornladios/ADIOS2/blob/master/testing/adios2/bindings/python/TestBPWriteTypesHighLevelAPI_HDF5.py

you can turn it uncommenting the hdf5 portions here:
https://github.com/ornladios/ADIOS2/blob/master/testing/adios2/bindings/python/CMakeLists.txt

Thanks!

**Comment by guj** [comment]
@pnorbert  Yes if a variable exists, the attribute is attached to it in the h5 file.

I could do the same for attributes. The reason I didn't is, without enforcing every node in the path should be valid,  conflicts would arise. E.g. someone first defineAttr("/a/b/c") and groups /a & /a/b are created with attr c under /a/b/.  Later they call defineVar("/a/b") This would be ok for adios but not ok for h5.  
I think IO shall restrict to use of defineattr(name, value) to only path that exists.

**Comment by williamfgc** [comment]
@guj  I saw a PR, please a couple of things: 
1) can we reference this issue on the PR commit (a simple #issue_number would do the trick)?
2) can we make sure the test I provided (I spent a lot of time on this) runs? I'm ok if you add your own tests, but every bug fix or new features must have tests (might not cover 100%) but some testing is better than none.
This is what we are all following. Thanks!

**Comment by guj** [comment]
Hi William,  Yes, your test runs.  I didn’t change the cmake file. I will do it with a another pull request

How do I add the issue_number to the PR? add it in the comment?

- Junmin

**Comment by williamfgc** [comment]
Type # and GitHub will try to autocomplete on it's desktop version as you type the issue number (no space)

**Comment by williamfgc** [comment]
you latest PR doesn't execute tests on this, CMake HDF5 test is still commented out....https://github.com/ornladios/ADIOS2/blob/master/testing/adios2/bindings/python/CMakeLists.txt

**Comment by guj** [comment]
I knew. Will do another PR later today - Junmin

**Comment by guj** [comment]
@williamfgc @pnorbert 
Hi, Here is something I need to clarify about Attribtues.

I was wrong on the previous post about path must exist in HDF5. This enforcement applies to vars  and groups, but not attributes.

So, after creating a new file.  I can do the following:
- create an int attribute "future/path" to the root directory"/" . 
- create a dataset "future",  then create a float attribute "path" to "future". 

when asking for attribute in HDF5, you need to specify the relative start point. 

Now I ask for attributes:
- (root, "future/path") gives me an int attribute
- ("future", "path") gives me a float attribute.

In ADIOS2, the above steps fail at the second create attribute. Because both cases are referred to as "future/path".  So the attribute to variable "future" can not exist. 

And ADIOS2 would return all the attributes with GetAttributesDataMap(). 
 
Is there a plan to add support to Attributes to have the same behavior as HDF5? 
Ideally would be:
1) it would be good to support the above scenario
2) whether 1) will be done or not, support GetAttributesDataMap(string base="") where base can be either root dir or a variable. This way ./bpls can list the attributes by variable to reflect the fact that some attrs are associated with a variable
3) will attributes be attached to steps (for variables) ? 
 
- Junmin

**Comment by williamfgc** [comment]
> In ADIOS2, the above steps fail at the second create attribute. Because both cases are referred to as "future/path". So the attribute to variable "future" can not exist.

This is a clear name conflict. I am not familiar how this is sorted out by HDF5. "future/path" attribute already exists. Is there a 3rd category besides unique variable names and unique attribute names?

>And ADIOS2 would return all the attributes with GetAttributesDataMap().

`AvailableAttributes` is the one doing the trick from your 2.

About 3. attributes are not "attached" to steps, you can redefined at every step using `RemoveAttribute` and `DefineAttribute`, but that scenario is better suited for a Variable (hence, the name "variable"). I am not sure if there is a use case for that using Attributes.

**Comment by guj** [comment]
@williamfgc  

Hi William, below is the display from HDF5:  (notice I created group "future" instead of dataset "future" in this file)  There are 2 attributes, 1 is called "future/attr" under "/"
the other is called "attr" under "future". In other words,  attributes are not identified by path. 
but rather groupname (or varname)  & attrname.

And by the way AvailbleAttributes(varname) helps getting attrs of a var. That is nice.  If varname is empty, does it return only top level attribtues (i.e. only global attributes, not including any attrs that are attached to a variable)?

--- H5 output showing 2 attributes -- 
h5dump  Attributes.h5 

HDF5 "Attributes.h5" {
GROUP "/" {
   ATTRIBUTE "future/attr" {
      DATATYPE  H5T_STD_I32LE
      DATASPACE  SCALAR
      DATA {
      (0): 123
      }
   }  
   GROUP "future" {
      ATTRIBUTE "attr" {
         DATATYPE  H5T_IEEE_F32LE
         DATASPACE  SCALAR
         DATA {
         (0): 9.123
         }
      }
   }
   DATASET "varA" {
      DATATYPE  H5T_STD_I32LE
      DATASPACE  SIMPLE { ( 7 ) / ( 7 ) }
      ….

**Comment by pnorbert** [comment]
Again, showing a weird example from a universally complete data model where
anything is possible. Can you also show how to list the attributes with
h5ls?
The only lesson I take from this: One should not use / in naming anything
in an HDF5 file.

**Comment by williamfgc** [comment]
> If varname is empty, does it return only top level attribtues (i.e. only global attributes, not including any attrs that are attached to a variable)?

As the function promises: it returns available attributes, not sure why the expectation is different. ADIOS doesn't provide a data model or a notion of global. It is the application sorting variables and attributes as needed in a conflict free flat hierarchy.

Why is required to support of case in which a group name = hierarchy delimiter "/"? The wonderful thing about adios is that you can pick up any string (not just /) as your hierarchy delimiter (e g :: , _, \\) so if your group is called "/" I'd use :: as your hierarchy delimiter.....that's still very obfuscated

**Comment by williamfgc** [comment]
An attribute can be "attached" to many variables, but not all in a dataset (example: "Using m/s for velocities")....does it make it global? Coming from working with physical simulation flows I always found silly the concept of global and variable attributes, what are those in between called? Partial?

**Comment by pnorbert** [comment]
@wfgodoy I am not sure if I follow your argument. ADIOS has global
attributes that are not attached to any variable. Question is how do
you get them.

**Comment by williamfgc** [comment]
@pnorbert we'll have to implement a loop through available variable names and remove those attributes that contain a variable name in position 0...at that point we are basically imposing a data model. My fear is that someone starts removing variables, would the attributes become dangling, global, etc.?  My point is more of a rant, don't pay too much attention. My experience with simulation codes is that they don't like being restricted by a library data model (we saw XGC had a different idea about namespaces and don't follow the UNIX/HDF5 convention).

**Comment by guj** [comment]
@pnorbert 
Well, lesson or not, I would like to clear up the semantics of attribute names that has a path. This affects how I load attr to HDF5. And if we support the HDF5 interface, it is necessart that we understand the differences in the two I/O systems.

So let me summarize:

0) attrs names can have seperators, and this path does not need to be valid  (yes-h5, yes-adios2)
1) attrs can be global or associated to a variable  (yes-h5, yes-adios2) 
2) attr names can overlap with a variable name (at the same hierarchy level) (yes-h5, yes-adios2)
3) can list all attrs associated with a variable  (yes-h5, yes-adios2)
4) can list all attrs not associated with any variable (yes-h5,  not_yet-adios2)
5) we do not identify attr name by its path,  (yes-h5, not_now-adios2) (see my example earlier today) 

So this is for @williamfgc 
If I do:
- tempVar= DefineVar("varTemp")
- attr0= DefineAttr("varA/unit")

I think attr0 will be considered associated to variable tempVar,  although it is not declared that way. 
In other words, InquireVar("unit", varname="varA") will return attr0.  (and is it supposed to be this way?)



- Junmin

**Comment by williamfgc** [comment]
@guj yes, because the assumption that there is a variable called "varA" is wrong in the first place. 
 In your example "varA/unit" is just a plain string that names an attribute. I think the limitations are given by HDF5 "/" to make us believe is a hierarchy delimiter. Would HDF5 provide namespace ("::", ".") or prefix ("_") style hierarchies?

**Comment by williamfgc** [comment]
> 4. can list all attrs not associated with any variable (yes-h5,  not_yet-adios2)

HDF5 can?

**Comment by williamfgc** [comment]
@guj BTW, I'd move the discussion somewhere else....let's focus on the current ticket and close it.

**Comment by guj** [comment]
@williamfgc: 

one way to do 4) is, 
- first call H5Oinfo(fileHandle) to get num attrs (fileHandle implies top level) . This function does not do recursive search. 
- then call H5Aget_name_by_idx(fileHandle)

about 5) bpls display attributes using a path.  For example,  "varA/unit". This is an absolute path. HDF5 has a different way for attributes and variables.  Attributes are referred as attr contained in a node. e.g. (nodeName, attrName) are needed to locate an attr.  Here attrName can be long, e.g. "no/such/path".  While Variables requires absolute path and has to be valid. 

In my enhanced example:
 
a) attr0= DefineAttr("varTemp/unit") // I typed varA before. sorry.
b) attr1 = DefineAttr("notVar/unit")
c) tempVar= DefineVar("varTemp")

attr1 is considered valid and belongs to no variables.
attr0 is not clear. 
 at step b) it does not belong to any variable
 at step c), it belongs to tempVar in ADIOS as of now, even though we did not define it that way

If this is how ADIOS attribute  behavior will be, I am ok with it.   

It would be useful though, to have an option to list only attrs in the IO at the top level, what notation we use for top level does not matter.  "/" or "." or ""   

I don't know whether HDF5 uses other seperators. 

Guess you can close the ticket now.

**Comment by williamfgc** [comment]
> Guess you can close the ticket now.

Is the provided test passing now?

**Comment by guj** [comment]
@williamfgc 
yes the python test passed when the PR was submitted. 
I didn't update the CMakefile in the PR. Will do it with the next one.

**Comment by williamfgc** [comment]
@guj OK, sounds good. That will turn it on for CI, after that we can close the ticket. Thanks!
